### PR TITLE
Make fetch test work with .onion domains

### PR DESCRIPTION
**Pull request by koops76**
Requires https://github.com/EFForg/https-everywhere-docker-base/pull/2 to work.

### CODE REVIEW

**Comment by koops76** [comment]
ping @Hainish.

**Comment by cschanaj** [comment]
did you try `./fetch-test.sh ./rules/FacebookCoreWWWi.onion.xml` locally?

File: test/rules/src/https_everywhere_checker/http_client.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 import traceback
 import subprocess
 import re
+from urlparse import urlparse
```
Comment (cschanaj) [inline test/rules/src/https_everywhere_checker/http_client.py]: see line 5

File: test/rules/src/https_everywhere_checker/http_client.py [hunk 2]
```diff
@@ -306,6 +307,8 @@ def staticFetch(url, options, platformPath):
 			buf = cStringIO.StringIO()
 			headerBuf = cStringIO.StringIO()
 			
+			urlParts = urlparse(url)
```
Comment (cschanaj) [inline test/rules/src/https_everywhere_checker/http_client.py]: maybe `urlParts = urlparse.urlparse(url)`?

**Comment by koops76** [comment]
@cschanaj No, for that I need @Hainish to update Docker base.

**Comment by koops76** [comment]
@cschanaj Done.

**Comment by cschanaj** [comment]
I get the following result on my local machine with tor installed. Not really sure why though. 

```
$ ./fetch-test.sh ./rules/FacebookCoreWWWi.onion.xml 
INFO Finished comparing http://developers.facebookcorewwwi.onion/ -> https://developers.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://m.facebookcorewwwi.onion/ -> https://m.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://www.facebookcorewwwi.onion/ -> https://www.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://facebookcorewwwi.onion/ -> https://facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://scontent.xx.fbcdn23dssr3jqnq.onion/ -> https://scontent.xx.fbcdn23dssr3jqnq.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://developers.facebookcorewwwi.onion/ => https://developers.facebookcorewwwi.onion/: CURLOPT_PROXY
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://m.facebookcorewwwi.onion/ => https://m.facebookcorewwwi.onion/: CURLOPT_PROXY
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://www.facebookcorewwwi.onion/ => https://www.facebookcorewwwi.onion/: CURLOPT_PROXY
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://facebookcorewwwi.onion/ => https://facebookcorewwwi.onion/: CURLOPT_PROXY
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://scontent.xx.fbcdn23dssr3jqnq.onion/ => https://scontent.xx.fbcdn23dssr3jqnq.onion/: CURLOPT_PROXY
INFO Finished in 0.01 seconds. Loaded rulesets: 1, URL pairs: 5.
```

**Comment by koops76** [comment]
@cschanaj Was Tor running?

**Comment by koops76** [comment]
@cschanaj Try `curl -v --socks5-hostname localhost:9050 https://facebookcorewwwi.onion`

**Comment by cschanaj** [comment]
`curl -v --socks5-hostname 127.0.0.1:9050 https://facebookcorewwwi.onion` works, I guess it is something related to pycurl

File: test/rules/src/https_everywhere_checker/http_client.py [hunk 3]
```diff
@@ -325,6 +327,9 @@ def staticFetch(url, options, platformPath):
 			c.setopt(c.SSLVERSION, options.sslVersion)
 			c.setopt(c.VERBOSE, options.curlVerbose)
 			c.setopt(c.SSL_CIPHER_LIST, options.cipherList)
+			if urlParts.hostname[-6:] == '.onion':
+				c.setopt(c.CURLOPT_PROXY, '127.0.0.1:9050')
```
Comment (cschanaj) [inline test/rules/src/https_everywhere_checker/http_client.py]: please replace line 331 and 332 with the following 

```python
c.setopt(c.PROXY, 'socks5://127.0.0.1:9050')
```
or 
```python 
c.setopt(c.PROXY, '127.0.0.1')
c.setopt(c.PROXYPORT, 9050)
c.setopt(c.PROXYTYPE, c.PROXYTYPE_SOCKS5_HOSTNAME)
```

this typo is so difficult to trace...

Comment (koops76) [inline test/rules/src/https_everywhere_checker/http_client.py]: @cschanaj Done. Working now?

**Comment by cschanaj** [comment]
@koops76 AFAIK, it works on my local machine and finished the `fetch-test` with exit 0. Please run your code at least locally next time, since `fetch-test` is not triggered by the core code changes.

**Comment by koops76** [comment]
Try testing a deliberately broken ruleset like replace the hostname with `doesntexistdoesntexist.onion`.

**Comment by cschanaj** [comment]
Not really sure if this is good enough. I'd like to leave this for @Hainish to decide. thanks.

```bash
$ ./fetch-test.sh ./rules/FacebookCoreWWWi.onion.xml INFO Finished comparing http://developers.facebookcorewwwi.onion/ -> https://developers.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://m.facebookcorewwwi.onion/ -> https://m.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://www.facebookcorewwwi.onion/ -> https://www.facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://doesntexistdoesn.onion/ -> https://doesntexistdoesn.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://facebookcorewwwi.onion/ -> https://facebookcorewwwi.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
INFO Finished comparing http://scontent.xx.fbcdn23dssr3jqnq.onion/ -> https://scontent.xx.fbcdn23dssr3jqnq.onion/. Rulefile: ./rules/FacebookCoreWWWi.onion.xml.
ERROR ./rules/FacebookCoreWWWi.onion.xml: Fetch error: http://doesntexistdoesn.onion/ => https://doesntexistdoesn.onion/: (7, "Can't complete SOCKS5 connection to 0.0.0.0:0. (4)")
INFO Finished in 63.31 seconds. Loaded rulesets: 1, URL pairs: 6.
```

P.S. I do not want to get too involved in this since I have other priorities in this project. Thanks for this PR anyway!!

**Comment by J0WI** [comment]
@koops76 the Docker base can be updated here: https://github.com/EFForg/https-everywhere-docker-base/blob/master/Dockerfile

**Comment by Hainish** [comment]
@koops76 you can test locally using docker.  Just run

```bash
sudo docker build -t electronicfrontierfoundation/https-everywhere-docker-base .
```

Once that image is built on your system, you can then run in this project

```bash
sudo docker build -t httpse .
sudo docker run --rm -ti -v $(pwd):/opt httpse bash -c "./fetch-test.sh ./rules/FacebookCoreWWWi.onion.xml"
```

**Comment by koops76** [comment]
@Hainish I'm not running Linux, so I can't test it.

**Comment by Hainish** [comment]
Thanks for this!

**Comment by koops76** [comment]
@Hainish See #11806.

**Pull request merged**